In [2]:
!pip install pandas
!pip install numpay

  Using cached pandas-2.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)
  Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
Using cached pandas-2.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.0 MB)
Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)
Using cached pytz-2024.1-py2.py3-none-any.whl (505 kB)
Using cached tzdata-2024.1-py2.py3-none-any.whl (345 kB)
ERROR: Could not find a version that satisfies the requirement numpay (from versions: none)
ERROR: No matching distribution found for numpay


In [352]:
import pandas as pd
import numpy as np
import random

df_ori = pd.read_csv('/home/oscar/escola/dados/Sales_Product_Combined.csv')

In [353]:
df_ori.head()

Order ID                     Product  Quantity Ordered  Price  Order Date  \
0    176558        USB-C Charging Cable                 2  11.95  19-04-2019   
1    176559  Bose SoundSport Headphones                 1  99.99  07-04-2019   
2    176560                Google Phone                 1    600  12-04-2019   
3    176560            Wired Headphones                 1  11.99  12-04-2019   
4    176561            Wired Headphones                 1  11.99  30-04-2019   

       Time                      Purchase Address          City Product Type  
0   8:46 AM          917 1st St, Dallas, TX 75001        Dallas        Cable  
1  10:30 PM     682 Chestnut St, Boston, MA 02215        Boston   Headphones  
2   2:38 PM  669 Spruce St, Los Angeles, CA 90001   Los Angeles        Phone  
3   2:38 PM  669 Spruce St, Los Angeles, CA 90001   Los Angeles   Headphones  
4   9:27 AM     333 8th St, Los Angeles, CA 90001   Los Angeles   Headphones

In [354]:
df_ori.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185950 entries, 0 to 185949
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          185950 non-null  int64 
 1   Product           185950 non-null  object
 2   Quantity Ordered  185950 non-null  int64 
 3   Price             185950 non-null  object
 4   Order Date        185950 non-null  object
 5   Time              185950 non-null  object
 6   Purchase Address  185950 non-null  object
 7   City              185950 non-null  object
 8   Product Type      185950 non-null  object
dtypes: int64(2), object(7)
memory usage: 12.8+ MB


# Teste 1

In [355]:
df = df_ori.rename(columns = {'Order ID': 'id', 'Product' : 'produto', 'Quantity Ordered': 'quant', 'Price': 'preco', 'Product Type': 'categoria', 'City': 'cidade'})
df['horario'] = pd.to_datetime(df['Order Date'] + ' ' + df['Time'], format='%d-%m-%Y %I:%M %p')
df['horario'] = df['horario'] + pd.DateOffset(hours = -6)
df['cidade'] = df['cidade'].str.strip()
df['rnd'] = [random.randrange(100) for k in df.index]
df = df[['id', 'horario', 'categoria', 'produto', 'quant', 'preco', 'cidade', 'rnd']].sort_values('horario')

In [356]:
df.head()

id             horario   categoria                   produto  \
73538  147268 2018-12-31 21:07:00  Headphones          Wired Headphones   
74345  148041 2018-12-31 21:40:00       Cable      USB-C Charging Cable   
75692  149343 2018-12-31 22:56:00  Headphones  Apple Airpods Headphones   
76343  149964 2018-12-31 23:53:00   Batteries    AAA Batteries (4-pack)   
75699  149350 2019-01-01 00:03:00       Cable      USB-C Charging Cable   

       quant  preco         cidade  rnd  
73538      1  11.99  New York City   44  
74345      1  11.95  San Francisco   37  
75692      1    150  New York City   21  
76343      1   2.99         Dallas   54  
75699      2  11.95        Atlanta    2

In [357]:
# existem registros fora do ano
# df.head()
df.iloc[-5:]
# df.query('horario < "2019-01-01"')
# df.query('horario > "2019-12-31 23:59:59"')

id             horario   categoria                     produto  \
39113  304165 2019-12-31 22:13:00   Batteries      AAA Batteries (4-pack)   
33863  299125 2019-12-31 22:21:00       Cable        USB-C Charging Cable   
40860  305840 2019-12-31 22:54:00  Headphones  Bose SoundSport Headphones   
35322  300519 2019-12-31 23:13:00  Headphones  Bose SoundSport Headphones   
35323  300519 2019-12-31 23:13:00       Cable    Lightning Charging Cable   

       quant  preco         cidade  rnd  
39113      1   2.99       Portland   87  
33863      1  11.95  New York City   96  
40860      1  99.99  San Francisco   54  
35322      1  99.99  New York City   37  
35323      1  14.95  New York City    6

In [358]:
# alterando datas de 2018 para 2019
for i in range(4):
  df.iloc[i, 1] = df.iloc[i, 1] + pd.DateOffset(hours = 3)

In [359]:
# deseja-se demonstrar um aumento de vendas nos meses 9, 10 e 11
b = df[['horario', 'quant']]
b.index = df.horario
b.resample(rule='ME', on='horario')['quant'].sum()

horario
2019-01-31    10873
2019-02-28    13464
2019-03-31    17003
2019-04-30    20569
2019-05-31    18654
2019-06-30    15252
2019-07-31    16076
2019-08-31    13436
2019-09-30    13133
2019-10-31    22694
2019-11-30    19816
2019-12-31    28109
Freq: ME, Name: quant, dtype: int64

In [360]:
# transferir 25% das vendas dos meses 3, 4 e 5 para os meses 9, 10 e 11
# df['horario'] = df.apply(
#   lambda row: row['horario'] + pd.DateOffset(months = 6) \
#     if random.randrange(4) < 1 and \
#       row['horario'] > pd.to_datetime('01-03-2019', format='%d-%m-%Y') and \
#       row['horario'] < pd.to_datetime('01-06-2019', format='%d-%m-%Y') \
#         else row['horario'], axis = 1)

df['horario'] = df['horario'].mask(
  (df['horario'] > pd.to_datetime('01-03-2019', format='%d-%m-%Y')) & 
  (df['horario'] < pd.to_datetime('01-06-2019', format='%d-%m-%Y')) & 
  (df['rnd'] < 26), 
  df['horario'] + pd.DateOffset(months = 6))

# transferir 20% das vendas do mês 12 para o mês 1
df['horario'] = df['horario'].mask(
  (df['horario'] > pd.to_datetime('01-12-2019', format='%d-%m-%Y')) & 
  (df['rnd'] < 20), 
  df['horario'] + pd.DateOffset(months = -11))

In [361]:
b = df[['horario', 'quant']]
b.index = df.horario
b.resample(rule='ME', on='horario')['quant'].sum()

horario
2019-01-31    16419
2019-02-28    13464
2019-03-31    12682
2019-04-30    15199
2019-05-31    13844
2019-06-30    15252
2019-07-31    16076
2019-08-31    13436
2019-09-30    17454
2019-10-31    28064
2019-11-30    24626
2019-12-31    22563
Freq: ME, Name: quant, dtype: int64

In [362]:
# df['horario'] = df.apply(lambda row: row['horario'] + pd.DateOffset(years = -random.randrange(10)), axis = 1)

# mask = df.horario.mask(df.cidade == 'Austin')
# df.loc[mask, 'horario'] = df['horario'] + pd.DateOffset(years = -5)

# df['horario'] = df['horario'].mask(df['cidade'] == 'Austin', other=lambda x: x * 2)
# df['horario'] = df['horario'].mask(df['cidade'] == 'Austin', other = lambda row: row + pd.DateOffset(years = -5))

# df[df['cidade'] == 'Austin']

df['horario'] = df['horario'].mask(df.cidade == 'Austin', df['horario'] + pd.DateOffset(years = -5))
df['horario'] = df['horario'].mask(df.cidade == 'Portland', df['horario'] + pd.DateOffset(years = -4))
df['horario'] = df['horario'].mask(df.cidade == 'Seattle', df['horario'] + pd.DateOffset(years = -3))
df['horario'] = df['horario'].mask(df.cidade == 'Atlanta', df['horario'] + pd.DateOffset(years = -2))
df['horario'] = df['horario'].mask(df.cidade == 'Dallas', df['horario'] + pd.DateOffset(years = -1))
df['horario'] = df['horario'].mask(df.cidade == 'New York City', df['horario'] + pd.DateOffset(years = 4))
df['horario'] = df['horario'].mask(
  (df.cidade == 'Los Angeles') &
  (df['rnd'] < 40), 
  df['horario'] + pd.DateOffset(years = 1))
df['horario'] = df['horario'].mask(
  (df.cidade == 'Los Angeles') &
  (df['rnd'] >= 40), 
  df['horario'] + pd.DateOffset(years = 2))

df['horario'] = df['horario'].mask(
  (df.cidade == 'San Francisco') &
  (df['rnd'] < 40), 
  df['horario'] + pd.DateOffset(years = 3))
df['horario'] = df['horario'].mask(
  (df.cidade == 'San Francisco') &
  (df['rnd'] >= 40), 
  df['horario'] + pd.DateOffset(years = 5))

# df

# x = {
# 'Austin': -5,
# 'Portland': -4,
# 'Seattle': -3,
# 'Atlanta': -2,
# 'Dallas': -1,
# 'Boston': 0,
# 'New York City': 4,
# 'Los Angeles': 0.4 * 1 + 0.6 * 2,
# 'San Francisco': 3 + 4,
# }


# random.randrange(5) < 2 ? 1 : 2

# Austin           11153
# Portland         14053
# Seattle          16553
# Atlanta          16602
# Dallas           16730
# Boston           22528
# New York City    27932
# Los Angeles      33289
# San Francisco    50239

# df['cidade'][:50]

In [278]:
b = df[['horario', 'quant']]
b.index = df.horario
b.query('horario < "2019-01-01"').resample(rule='ME', on='horario')['quant'].sum()

horario
2014-01-31     881
2014-02-28     685
2014-03-31     710
2014-04-30     819
2014-05-31     678
2014-06-30     804
2014-07-31     848
2014-08-31     735
2014-09-30    1006
2014-10-31    1506
2014-11-30    1260
2014-12-31    1221
2015-01-31    1035
2015-02-28     938
2015-03-31     824
2015-04-30    1055
2015-05-31     953
2015-06-30    1081
2015-07-31    1027
2015-08-31     928
2015-09-30    1125
2015-10-31    1897
2015-11-30    1677
2015-12-31    1513
2016-01-31    1273
2016-02-29    1066
2016-03-31     972
2016-04-30    1198
2016-05-31    1021
2016-06-30    1198
2016-07-31    1265
2016-08-31    1148
2016-09-30    1422
2016-10-31    2288
2016-11-30    1965
2016-12-31    1737
2017-01-31    1316
2017-02-28    1043
2017-03-31    1011
2017-04-30    1231
2017-05-31    1115
2017-06-30    1224
2017-07-31    1259
2017-08-31    1053
2017-09-30    1408
2017-10-31    2199
2017-11-30    1958
2017-12-31    1785
2018-01-31    1334
2018-02-28    1108
2018-03-31     999
2018-04-30    1157
2018

In [279]:
b = df[['horario', 'quant']]
b.index = df.horario
b.query('horario > "2019-01-01" and horario < "2024-01-01"').resample(rule='ME', on='horario')['quant'].sum()

horario
2019-01-31    1814
2019-02-28    1403
2019-03-31    1378
2019-04-30    1595
2019-05-31    1465
2019-06-30    1616
2019-07-31    1771
2019-08-31    1409
2019-09-30    1959
2019-10-31    3041
2019-11-30    2692
2019-12-31    2385
2020-01-31    1567
2020-02-29     878
2020-03-31     350
2020-04-30     477
2020-05-31     394
2020-06-30     978
2020-07-31     992
2020-08-31     825
2020-09-30    1636
2020-10-31    2367
2020-11-30    1987
2020-12-31     895
2021-01-31    1023
2021-02-28    1283
2021-03-31    1650
2021-04-30    2058
2021-05-31    1733
2021-06-30    1472
2021-07-31    1515
2021-08-31    1245
2021-09-30    1247
2021-10-31    2216
2021-11-30    1897
2021-12-31    2604
2022-01-31    2371
2022-02-28    1308
2022-03-31     550
2022-04-30     673
2022-05-31     639
2022-06-30    1536
2022-07-31    1580
2022-08-31    1236
2022-09-30    2317
2022-10-31    3361
2022-11-30    3140
2022-12-31    1342
2023-01-31    2249
2023-02-28    1808
2023-03-31    1645
2023-04-30    2011
2023

In [288]:
b = df[['horario', 'quant']]
b.index = df.horario
b.query('horario > "2020-01-01"').resample(rule='ME', on='horario')['quant'].sum()

horario
2020-01-31    1536
2020-02-29     911
2020-03-31     429
2020-04-30     459
2020-05-31     417
2020-06-30     961
2020-07-31     997
2020-08-31     857
2020-09-30    1552
2020-10-31    2415
2020-11-30    1999
2020-12-31     904
2021-01-31    1021
2021-02-28    1250
2021-03-31    1594
2021-04-30    2065
2021-05-31    1734
2021-06-30    1489
2021-07-31    1510
2021-08-31    1213
2021-09-30    1308
2021-10-31    2179
2021-11-30    1861
2021-12-31    2628
2022-01-31    2413
2022-02-28    1326
2022-03-31     573
2022-04-30     654
2022-05-31     647
2022-06-30    1461
2022-07-31    1617
2022-08-31    1259
2022-09-30    2252
2022-10-31    3457
2022-11-30    3028
2022-12-31    1290
2023-01-31    2294
2023-02-28    1808
2023-03-31    1617
2023-04-30    2045
2023-05-31    1880
2023-06-30    2000
2023-07-31    2169
2023-08-31    1804
2023-09-30    2310
2023-10-31    3651
2023-11-30    3327
2023-12-31    3027
2024-01-31    1568
2024-02-29    1926
2024-03-31    2430
2024-04-30    3022
2024

In [364]:
df.groupby('cidade')['quant'].count().sort_values()

cidade
Austin            9905
Portland         12465
Seattle          14732
Dallas           14820
Atlanta          14881
Boston           19934
New York City    24876
Los Angeles      29605
San Francisco    44732
Name: quant, dtype: int64

In [17]:
# df.reset_index(drop = True, inplace = True)
# df['categoria', 'produto'].unique()
# list(df.categoria.unique())
df.groupby(['categoria', 'produto']).size()


categoria   produto                   
Batteries   AA Batteries (4-pack)         20577
            AAA Batteries (4-pack)        20641
Cable       Lightning Charging Cable      21658
            USB-C Charging Cable          21903
Dryer       LG Dryer                        646
Headphones  Apple Airpods Headphones      15549
            Bose SoundSport Headphones    13325
            Wired Headphones              18882
Laptop      Macbook Pro Laptop             4724
            ThinkPad Laptop                4128
Machine     LG Washing Machine              666
Monitor     20in Monitor                   4101
            27in 4K Gaming Monitor         6230
            27in FHD Monitor               7507
            34in Ultrawide Monitor         6181
Phone       Google Phone                   5525
            Vareebadd Phone                2065
TV          Flatscreen TV                  4800
iPhone      iPhone                         6842
dtype: int64

Product Type  Product                   
Batteries     AA Batteries (4-pack)             3.84
              AAA Batteries (4-pack)            2.99
Cable         Lightning Charging Cable         14.95
              USB-C Charging Cable             11.95
Dryer         LG Dryer                           600
Headphones    Apple Airpods Headphones           150
              Bose SoundSport Headphones       99.99
              Wired Headphones                 11.99
Laptop        Macbook Pro Laptop            1,700.00
              ThinkPad Laptop                 999.99
Machine       LG Washing Machine                 600
Monitor       20in Monitor                    109.99
              27in 4K Gaming Monitor          389.99
              27in FHD Monitor                149.99
              34in Ultrawide Monitor          379.99
Phone         Google Phone                       600
              Vareebadd Phone                    400
TV            Flatscreen TV                      300
iPhone        iPhone                             700
Name: Price, dtype: object



categoria  produto
Cat01      Prod01       38.40
           Prod02       29.90
Cat02      Prod03      149.50
           Prod04      119.50
Cat03      Prod05      600.00
Cat04      Prod06      150.00
           Prod07       99.99
           Prod08      119.90
Cat05      Prod09      700.00
Cat06      Prod10     1700.00
           Prod11      999.99
Cat07      Prod12      600.00
Cat08      Prod13      109.99
           Prod14      389.99
           Prod15      149.99
           Prod16      379.99
Cat09      Prod17      600.00
           Prod18      400.00
Cat10      Prod79      300.00
Name: preco, dtype: float64




Cat01  Xicaras
Prod01 LEITEIRA TEXAS MEDIA 550ML por R$ 29,85
Prod02 XICARA CAFÉ C/ PIRES VENEZA por R$ 14,69

Cat02  Biscoitos, Bolos e Torradas em Sache
Prod03 SACHE BISCOITO AMANTEIGADO COCO RENATA SACHÊ 9G 2X2 CAIXA 280 UND por R$ 87,87
Prod04 SACHE DE BISCOITO MARIA RENATA 11G CAIXA 180 UNIDADES por R$ 68,20

Cat03  Toalha Personalizada Para Hotelaria
Prod05 TOALHA DE BANHO BORDADA TURQUESA CX 10 UND por R$ 544,23

Cat04  Capas Protetoras
Prod06 CAPA PROTETORA DE TRAVESSEIRO SCOTHGARD C/ZIPER- REPELENTE A ÁGUA CX 10 UND por R$ 212,69
Prod07 CAPA PROTETORA DE TRAVESSEIRO LAMINADO JADE - CX 10 UND por R$ 121,16
Prod08 CAPA PROTETORA COLCHÃO IMPERMEÁVEL LAMINADO SLEEP JADE A partir de R$ 47,43

Cat05  Mini Sabonete Para Hotel
Prod09 MINI SABONETE HOTEL REDONDO NATURALLYS PLISSADO 20G CX 500 UND por R$ 563,31

Cat06  Colchao de Hotel Mola
Prod10 COLCHÃO MOLA PARAHOTEL MAXI QUEEN 158X198X27 por R$ 2.035,00
Prod11 COLCHÃO MOLA PARAHOTEL MAXI SOLTEIRO 78X188X27 COM VELCRO por R$ 1.197,00

Cat07  Aromatizante Ar Condicionado
Prod12 PASTILHA BACTERIDA E AROMATIZANTE AR CONDICIONADO CX 10 UND por R$ 161,40

Cat08  Sabonetes e Amenities Para Hotelaria
Prod13 PENTE PLASTICO PERSONALIZADO PARA MOTEL DESCARTAVEL - CX 100 UND por R$ 107,00
Prod14 KIT SAIS DE BANHO PERSONALIZADO 40 G - SORTIDO CX 100 UND por R$ 430,00
Prod15 ESPUMA DE BANHO PERSONALIZADO EM FRASCO 30 ML - CX 100 UND por R$ 173,29
Prod16 KIT AMENITIES PERSONALIZADO SABONETE 15 G E SHAMPOO 30 ML HOTEL 100 UND por R$ 342,95

Cat09  Lençol de Hotel, Cobertores e AFins
Prod17 LENÇOL PERCAL 180 FIOS LINHA ESMERALDA 60/40 SUPER QUEEN CX 10 UND por R$ 1.031,00
Prod18 LENÇOL PERCAL 180 FIOS LINHA ESMERALDA 60/40 QUEEN CX 10 UND por R$ 963,00

Cat10  Toalha Para Hotel
Prod79 KIT 20 TOALHAS DE ROSTO PARA HOTEL, MOTEL, POUSADA por R$ 353,00



In [39]:
# df.to_csv('/home/oscar/escola/dados/saida.csv')
exp = df[['horario', 'categoria', 'produto', 'quant', 'preco', 'rnd']].query('horario > "2024-01-01"').sort_values('horario')
exp.to_csv('/home/oscar/escola/dados/saida2.csv', encoding='latin-1', index = False)

# Teste 2

In [344]:
df = df_ori.rename(columns = {'Order ID': 'id', 'Product' : 'produto', 'Quantity Ordered': 'quant', 'Price': 'preco', 'Product Type': 'categoria', 'City': 'cidade'})


In [345]:
df.groupby(['categoria', 'produto'])['preco'].max()

categoria   produto                   
Batteries   AA Batteries (4-pack)             3.84
            AAA Batteries (4-pack)            2.99
Cable       Lightning Charging Cable         14.95
            USB-C Charging Cable             11.95
Dryer       LG Dryer                           600
Headphones  Apple Airpods Headphones           150
            Bose SoundSport Headphones       99.99
            Wired Headphones                 11.99
Laptop      Macbook Pro Laptop            1,700.00
            ThinkPad Laptop                 999.99
Machine     LG Washing Machine                 600
Monitor     20in Monitor                    109.99
            27in 4K Gaming Monitor          389.99
            27in FHD Monitor                149.99
            34in Ultrawide Monitor          379.99
Phone       Google Phone                       600
            Vareebadd Phone                    400
TV          Flatscreen TV                      300
iPhone      iPhone                         

In [346]:
df['horario'] = pd.to_datetime(df['Order Date'] + ' ' + df['Time'], format='%d-%m-%Y %I:%M %p')
df['horario'] = df['horario'] + pd.DateOffset(hours = -6)
df['cidade'] = df['cidade'].str.strip()
df['preco'] = df['preco'].str.replace(',', '').astype(float)
# df['preco2'] = df['preco'] * (10 / (10 * df['preco'] / 400 + 1).astype(int) + 1).astype(int)


In [347]:
df = df[['id', 'horario', 'categoria', 'produto', 'quant', 'preco', 'cidade', 'rnd']].query('cidade == "San Francisco"').sort_values('horario')

In [348]:
cat = {
  'Batteries': 'Cat01',
  'Cable': 'Cat02',
  'Dryer': 'Cat03',
  'Headphones': 'Cat04',
  'iPhone': 'Cat05',
  'Laptop': 'Cat06',
  'Machine': 'Cat07',
  'Monitor': 'Cat08',
  'Phone': 'Cat09',
  'TV': 'Cat10',
}

prod = {
'20in Monitor': 'Prod01',
'27in 4K Gaming Monitor': 'Prod02',
'27in FHD Monitor': 'Prod03',
'34in Ultrawide Monitor': 'Prod04',
'AA Batteries (4-pack)': 'Prod05',
'AAA Batteries (4-pack)': 'Prod06',
'Apple Airpods Headphones': 'Prod07',
'Bose SoundSport Headphones': 'Prod08',
'Flatscreen TV': 'Prod09',
'Google Phone': 'Prod10',
'LG Dryer': 'Prod11',
'LG Washing Machine': 'Prod12',
'Lightning Charging Cable': 'Prod13',
'Macbook Pro Laptop': 'Prod14',
'ThinkPad Laptop': 'Prod15',
'USB-C Charging Cable': 'Prod16',
'Vareebadd Phone': 'Prod17',
'Wired Headphones': 'Prod18',
'iPhone': 'Prod19',
}

df['categoria'] = df.apply(lambda row: cat[row['categoria']], axis = 1)
df['produto'] = df.apply(lambda row: prod[row['produto']], axis = 1)

df['preco2'] = df.apply(lambda row: row['preco'] * 10 if row['preco'] < 80 else row['preco'], axis = 1)


In [75]:
df.groupby(['produto'])['preco'].max()

produto
Prod01      109.99
Prod02      389.99
Prod03      149.99
Prod04      379.99
Prod05        3.84
Prod06        2.99
Prod07         150
Prod08       99.99
Prod09         300
Prod10         600
Prod11         600
Prod12         600
Prod13       14.95
Prod14    1,700.00
Prod15      999.99
Prod16       11.95
Prod17         400
Prod18       11.99
Prod19         700
Name: preco, dtype: object

In [76]:
df.groupby(['categoria', 'produto'])['preco'].max()

categoria  produto
Cat01      Prod05         3.84
           Prod06         2.99
Cat02      Prod13        14.95
           Prod16        11.95
Cat03      Prod11          600
Cat04      Prod07          150
           Prod08        99.99
           Prod18        11.99
Cat05      Prod19          700
Cat06      Prod14     1,700.00
           Prod15       999.99
Cat07      Prod12          600
Cat08      Prod01       109.99
           Prod02       389.99
           Prod03       149.99
           Prod04       379.99
Cat09      Prod10          600
           Prod17          400
Cat10      Prod09          300
Name: preco, dtype: object

In [349]:
df['horario'] = df['horario'].mask(
  (df.rnd < 25), 
  df['horario'] + pd.DateOffset(years = 2))
df['horario'] = df['horario'].mask(
  (df.rnd >= 25) & 
  (df.rnd < 50),
  df['horario'] + pd.DateOffset(years = 3))
df['horario'] = df['horario'].mask(
  (df.rnd >= 50) & 
  (df.rnd < 75),
  df['horario'] + pd.DateOffset(years = 4))
df['horario'] = df['horario'].mask(
  (df.rnd >= 75),
  df['horario'] + pd.DateOffset(years = 5))

In [350]:
df['data'] = pd.to_datetime(df['horario']).dt.date
df['nota'] = df['horario'].dt.strftime('%Y%m%d-%M').str.slice(stop = 10)
# .apply(str).str.slice(stop = 3)

df

id             horario categoria produto  quant   preco  \
74345  148041 2021-12-31 21:40:00     Cat02  Prod16      6   11.95   
75984  149620 2022-01-01 00:34:00     Cat02  Prod13      8   14.95   
72260  146039 2022-01-01 01:24:00     Cat08  Prod04      1  379.99   
74395  148088 2021-01-01 01:37:00     Cat04  Prod18     11   11.99   
75947  149586 2021-01-01 01:47:00     Cat02  Prod16      3   11.95   
...       ...                 ...       ...     ...    ...     ...   
45255  310074 2023-12-31 19:46:00     Cat04  Prod07     11  150.00   
49912  314552 2022-12-31 19:47:00     Cat01  Prod06      7    2.99   
47264  312012 2021-12-31 20:12:00     Cat08  Prod03      1  149.99   
43292  308185 2022-12-31 22:06:00     Cat08  Prod03      8  149.99   
40860  305840 2021-12-31 22:54:00     Cat04  Prod08      8   99.99   

              cidade  rnd  preco2        data        nota  
74345  San Francisco   37  119.50  2021-12-31  20211231-4  
75984  San Francisco   31  149.50  2022-01-01  20220101-3  
72260  San Francisco   26  379.99  2022-01-01  20220101-2  
74395  San Francisco   22  119.90  2021-01-01  20210101-3  
75947  San Francisco    4  119.50  2021-01-01  20210101-4  
...              ...  ...     ...         ...         ...  
45255  San Francisco   72  150.00  2023-12-31  20231231-4  
49912  San Francisco   30   29.90  2022-12-31  20221231-4  
47264  San Francisco    2  149.99  2021-12-31  20211231-1  
43292  San Francisco   28  149.99  2022-12-31  20221231-0  
40860  San Francisco    6   99.99  2021-12-31  20211231-5  

[44732 rows x 11 columns]

In [351]:
df['quant2'] = (df['quant'] / (df['quant'] * df['preco'] / 10000 + 1).astype(int)).astype(int)
# df.query('horario > "2024-02-01"').sort_values(['data', 'nota']).head(25)
df.query('horario > "2024-02-01"').sort_values(['data', 'nota']).head(26)
# df.sort_values(['data', 'nota']).head(25)

id             horario categoria produto  quant    preco  \
64749  159630 2024-02-01 04:04:00     Cat02  Prod16     13    11.95   
65283  160149 2024-02-01 16:02:00     Cat02  Prod16      7    11.95   
61600  156604 2024-02-01 19:00:00     Cat08  Prod02      5   389.99   
63599  158520 2024-02-01 06:19:00     Cat02  Prod16     10    11.95   
64243  159142 2024-02-01 15:17:00     Cat02  Prod16      3    11.95   
64208  159108 2024-02-01 18:17:00     Cat06  Prod14      7  1700.00   
57109  152295 2024-02-01 04:23:00     Cat04  Prod08     15    99.99   
57234  152418 2024-02-01 04:24:00     Cat08  Prod04     11   379.99   
55451  150710 2024-02-01 04:27:00     Cat08  Prod02     23   389.99   
67110  161904 2024-02-01 05:29:00     Cat08  Prod03     14   149.99   
59152  154258 2024-02-01 07:29:00     Cat02  Prod13      2    14.95   
57628  152798 2024-02-01 11:27:00     Cat01  Prod05      9     3.84   
58266  153408 2024-02-01 12:29:00     Cat04  Prod08     12    99.99   
65040  159911 2024-02-01 16:20:00     Cat02  Prod13     19    14.95   
56078  151307 2024-02-01 16:27:00     Cat08  Prod01     15   109.99   
59082  154189 2024-02-01 06:31:00     Cat09  Prod10      4   600.00   
55310  150576 2024-02-01 19:33:00     Cat04  Prod08     11    99.99   
60657  155700 2024-02-01 09:43:00     Cat02  Prod16      9    11.95   
66350  161177 2024-02-01 11:45:00     Cat04  Prod08     14    99.99   
63967  158878 2024-02-01 01:54:00     Cat08  Prod02     26   389.99   
60082  155150 2024-02-01 01:57:00     Cat04  Prod07     13   150.00   
66377  161203 2024-02-01 11:50:00     Cat02  Prod13      4    14.95   
62339  157309 2024-02-01 11:57:00     Cat01  Prod05     16     3.84   
64369  159260 2024-02-01 13:54:00     Cat02  Prod13      2    14.95   
58655  153783 2024-02-01 14:53:00     Cat09  Prod17      1   400.00   
58283  153425 2024-02-02 01:04:00     Cat02  Prod13      1    14.95   

              cidade  rnd   preco2        data        nota  quant2  
64749  San Francisco   85   119.50  2024-02-01  20240201-0      13  
65283  San Francisco   75   119.50  2024-02-01  20240201-0       7  
61600  San Francisco   77   389.99  2024-02-01  20240201-0       5  
63599  San Francisco   99   119.50  2024-02-01  20240201-1      10  
64243  San Francisco   95   119.50  2024-02-01  20240201-1       3  
64208  San Francisco   85  1700.00  2024-02-01  20240201-1       3  
57109  San Francisco   79    99.99  2024-02-01  20240201-2      15  
57234  San Francisco   85   379.99  2024-02-01  20240201-2      11  
55451  San Francisco   86   389.99  2024-02-01  20240201-2      23  
67110  San Francisco   90   149.99  2024-02-01  20240201-2      14  
59152  San Francisco   87   149.50  2024-02-01  20240201-2       2  
57628  San Francisco   78    38.40  2024-02-01  20240201-2       9  
58266  San Francisco   86    99.99  2024-02-01  20240201-2      12  
65040  San Francisco   79   149.50  2024-02-01  20240201-2      19  
56078  San Francisco   81   109.99  2024-02-01  20240201-2      15  
59082  San Francisco   86   600.00  2024-02-01  20240201-3       4  
55310  San Francisco   95    99.99  2024-02-01  20240201-3      11  
60657  San Francisco   77   119.50  2024-02-01  20240201-4       9  
66350  San Francisco   99    99.99  2024-02-01  20240201-4      14  
63967  San Francisco   75   389.99  2024-02-01  20240201-5      13  
60082  San Francisco   96   150.00  2024-02-01  20240201-5      13  
66377  San Francisco   98   149.50  2024-02-01  20240201-5       4  
62339  San Francisco   89    38.40  2024-02-01  20240201-5      16  
64369  San Francisco   99   149.50  2024-02-01  20240201-5       2  
58655  San Francisco   78   400.00  2024-02-01  20240201-5       1  
58283  San Francisco   94   149.50  2024-02-02  20240202-0       1

In [90]:
exp = df[['data', 'horario', 'categoria', 'produto', 'quant', 'preco']].sort_values('horario')
exp.to_csv('/home/oscar/escola/dados/saida3.csv', index = False)

# Teste 3

In [ ]:
df_ori = pd.read_csv('/home/oscar/escola/dados/Sales_Product_Combined.csv')

In [ ]:
df_ori

Order ID                     Product  Quantity Ordered   Price  \
0         176558        USB-C Charging Cable                 2   11.95   
1         176559  Bose SoundSport Headphones                 1   99.99   
2         176560                Google Phone                 1     600   
3         176560            Wired Headphones                 1   11.99   
4         176561            Wired Headphones                 1   11.99   
...          ...                         ...               ...     ...   
185945    259353      AAA Batteries (4-pack)                 3    2.99   
185946    259354                      iPhone                 1     700   
185947    259355                      iPhone                 1     700   
185948    259356      34in Ultrawide Monitor                 1  379.99   
185949    259357        USB-C Charging Cable                 1   11.95   

        Order Date      Time                         Purchase Address  \
0       19-04-2019   8:46 AM             917 1st St, Dallas, TX 75001   
1       07-04-2019  10:30 PM        682 Chestnut St, Boston, MA 02215   
2       12-04-2019   2:38 PM     669 Spruce St, Los Angeles, CA 90001   
3       12-04-2019   2:38 PM     669 Spruce St, Los Angeles, CA 90001   
4       30-04-2019   9:27 AM        333 8th St, Los Angeles, CA 90001   
...            ...       ...                                      ...   
185945  17-09-2019   8:56 PM   840 Highland St, Los Angeles, CA 90001   
185946  01-09-2019   4:00 PM  216 Dogwood St, San Francisco, CA 94016   
185947  23-09-2019   7:39 AM     220 12th St, San Francisco, CA 94016   
185948  19-09-2019   5:30 PM   511 Forest St, San Francisco, CA 94016   
185949  30-09-2019  12:18 AM   250 Meadow St, San Francisco, CA 94016   

                  City Product Type  
0               Dallas        Cable  
1               Boston   Headphones  
2          Los Angeles        Phone  
3          Los Angeles   Headphones  
4          Los Angeles   Headphones  
...                ...          ...  
185945     Los Angeles    Batteries  
185946   San Francisco       iPhone  
185947   San Francisco       iPhone  
185948   San Francisco      Monitor  
185949   San Francisco        Cable  

[185950 rows x 9 columns]

In [ ]:
# valores originais
df_ori.groupby(['Product Type', 'Product'])['Price'].max()

Product Type  Product                   
Batteries     AA Batteries (4-pack)             3.84
              AAA Batteries (4-pack)            2.99
Cable         Lightning Charging Cable         14.95
              USB-C Charging Cable             11.95
Dryer         LG Dryer                           600
Headphones    Apple Airpods Headphones           150
              Bose SoundSport Headphones       99.99
              Wired Headphones                 11.99
Laptop        Macbook Pro Laptop            1,700.00
              ThinkPad Laptop                 999.99
Machine       LG Washing Machine                 600
Monitor       20in Monitor                    109.99
              27in 4K Gaming Monitor          389.99
              27in FHD Monitor                149.99
              34in Ultrawide Monitor          379.99
Phone         Google Phone                       600
              Vareebadd Phone                    400
TV            Flatscreen TV                      300
iPhon

In [ ]:
# renomeando campos
df_aux = df_ori.rename(columns={
  'Order ID': 'id', 
  'Product' : 'produto', 
  'Quantity Ordered': 'quantidade', 
  'Price': 'preco', 
  'Order Date': 'data',
  'Time': 'hora',
  'Product Type': 'categoria', 
  'Purchase Address': 'endereco', 
  'City': 'cidade',
})

# criando campo horário
df_aux['horario'] = pd.to_datetime(df_aux['data'] + ' ' + df_aux['hora'], format='%d-%m-%Y %I:%M %p')

# diminuindo quantidade de registros
df_aux['cidade'] = df_aux['cidade'].str.strip()
df_aux = df_aux.query('horario.dt.year == 2019 and cidade == "San Francisco"')[
  ['horario', 'categoria', 'produto', 'quantidade', 'preco']
]

In [ ]:
df = df_aux.sort_values('horario')

# campo auxiliar para aleatórios
df['rnd'] = [random.randrange(100) for _ in df.index]

# transferir 25% das vendas dos meses 3, 4 e 5 para os meses 9, 10 e 11
df['horario'] = df['horario'].mask(
  (df['horario'] > pd.to_datetime('01-03-2019', format='%d-%m-%Y')) & 
  (df['horario'] < pd.to_datetime('01-06-2019', format='%d-%m-%Y')) & 
  (df['rnd'] < 25), 
  df['horario'] + pd.DateOffset(months=6))

# transferir 20% das vendas do mês 12 para o mês 1
df['horario'] = df['horario'].mask(
  (df['horario'] > pd.to_datetime('01-12-2019', format='%d-%m-%Y')) & 
  (df['rnd'] < 20), 
  df['horario'] + pd.DateOffset(months=-11))

# distribuir aleatóriamente os registros em 5 anos
df['rnd'] = [random.randrange(100) for _ in df.index]
df['horario'] = df['horario'].mask(
  (df.rnd < 25), 
  df['horario'] + pd.DateOffset(years=2))
df['horario'] = df['horario'].mask(
  (df.rnd >= 25) & 
  (df.rnd < 50),
  df['horario'] + pd.DateOffset(years=3))
df['horario'] = df['horario'].mask(
  (df.rnd >= 50) & 
  (df.rnd < 75),
  df['horario'] + pd.DateOffset(years=4))
df['horario'] = df['horario'].mask(
  (df.rnd >= 75),
  df['horario'] + pd.DateOffset(years=5))

# campo auxiliar de data
df['data'] = pd.to_datetime(df['horario']).dt.date

# distribuir os itens do pedido
df['nota'] = df['horario'].dt.strftime('%Y%m%d-%M').str.slice(stop=10)

# permitir cálculo com o preço
df['preco'] = df['preco'].str.replace(',', '').astype(float)
df['preco'] = df.apply(lambda row: row['preco'] * 10 if row['preco'] < 80 else row['preco'], axis=1)
df['preco'] = df['preco'].round(2)

# calcular aleatóriamente quantidade, mas valores menores tem peso maior
pop = [i + 1 for i in range(40)]
wei = [i * i + 1 for i in range(40, 0, -1)]
df['quantidade'] = random.choices(pop, wei, k=len(df.index))
df['quantidade'] = (df['quantidade'] / (df['quantidade'] * df['preco'] / 10000 + 1).astype(int)).astype(int)
df['total'] = df['quantidade'] * df['preco']
df['total'] = df['total'].round(2)

# renomear categorias e produtos
cat = {
  'Batteries': 'Cat01',
  'Cable': 'Cat02',
  'Dryer': 'Cat03',
  'Headphones': 'Cat04',
  'iPhone': 'Cat05',
  'Laptop': 'Cat06',
  'Machine': 'Cat07',
  'Monitor': 'Cat08',
  'Phone': 'Cat09',
  'TV': 'Cat10',
}
prod = {
  'AA Batteries (4-pack)': 'Prod01',
  'AAA Batteries (4-pack)': 'Prod02',
  'Lightning Charging Cable': 'Prod03',
  'USB-C Charging Cable': 'Prod04',
  'LG Dryer': 'Prod05',
  'Apple Airpods Headphones': 'Prod06',
  'Bose SoundSport Headphones': 'Prod07',
  'Wired Headphones': 'Prod08',
  'iPhone': 'Prod09',
  'Macbook Pro Laptop': 'Prod10',
  'ThinkPad Laptop': 'Prod11',
  'LG Washing Machine': 'Prod12',
  '20in Monitor': 'Prod13',
  '27in 4K Gaming Monitor': 'Prod14',
  '27in FHD Monitor': 'Prod15',
  '34in Ultrawide Monitor': 'Prod16',
  'Google Phone': 'Prod17',
  'Vareebadd Phone': 'Prod18',
  'Flatscreen TV': 'Prod79',
}
df['categoria'] = df.apply(lambda row: cat[row['categoria']], axis=1)
df['produto'] = df.apply(lambda row: prod[row['produto']], axis=1)

df = df[['nota', 'data', 'horario', 'categoria', 'produto', 'quantidade', 'preco', 'total']].sort_values('horario')

In [ ]:
df.groupby(['categoria', 'produto'])['preco'].max()

categoria  produto
Cat01      Prod01       38.40
           Prod02       29.90
Cat02      Prod03      149.50
           Prod04      119.50
Cat03      Prod05      600.00
Cat04      Prod06      150.00
           Prod07       99.99
           Prod08      119.90
Cat05      Prod09      700.00
Cat06      Prod10     1700.00
           Prod11      999.99
Cat07      Prod12      600.00
Cat08      Prod13      109.99
           Prod14      389.99
           Prod15      149.99
           Prod16      379.99
Cat09      Prod17      600.00
           Prod18      400.00
Cat10      Prod79      300.00
Name: preco, dtype: float64

In [ ]:
df.head()

nota        data             horario categoria produto  \
75984  20210101-3  2021-01-01 2021-01-01 06:34:00     Cat02  Prod03   
74395  20210101-3  2021-01-01 2021-01-01 07:37:00     Cat04  Prod08   
75947  20210101-4  2021-01-01 2021-01-01 07:47:00     Cat02  Prod04   
76707  20210101-2  2021-01-01 2021-01-01 14:25:00     Cat10  Prod79   
38349  20210101-4  2021-01-01 2021-01-01 14:46:00     Cat01  Prod02   

       quantidade  preco   total  
75984           7  149.5  1046.5  
74395           2  119.9   239.8  
75947           6  119.5   717.0  
76707          27  300.0  8100.0  
38349          25   29.9   747.5

In [ ]:
df.info()
# df[:20]
# df[-20:]
# df.shape
# df_aux.shape
# df_aux[:20]
# df_aux[-20:]

<class 'pandas.core.frame.DataFrame'>
Index: 44720 entries, 74395 to 52516
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   nota        44720 non-null  object        
 1   data        44720 non-null  object        
 2   horario     44720 non-null  datetime64[ns]
 3   categoria   44720 non-null  object        
 4   produto     44720 non-null  object        
 5   quantidade  44720 non-null  int64         
 6   preco       44720 non-null  float64       
 7   total       44720 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 3.1+ MB


In [ ]:
b = df[['horario', 'quantidade']]
b.index = df.horario
b.resample(rule='ME', on='horario')['quantidade'].sum()

horario
2021-01-31     8683
2021-02-28     7523
2021-03-31     6928
2021-04-30     8009
2021-05-31     7406
2021-06-30     8660
2021-07-31     8721
2021-08-31     7426
2021-09-30     9598
2021-10-31    14900
2021-11-30    13538
2021-12-31    12011
2022-01-31     8733
2022-02-28     7707
2022-03-31     6737
2022-04-30     8590
2022-05-31     7817
2022-06-30     8341
2022-07-31     9022
2022-08-31     7018
2022-09-30     9231
2022-10-31    14920
2022-11-30    13717
2022-12-31    11533
2023-01-31     9305
2023-02-28     6768
2023-03-31     6613
2023-04-30     8364
2023-05-31     7922
2023-06-30     7763
2023-07-31     9282
2023-08-31     6977
2023-09-30     8662
2023-10-31    14771
2023-11-30    13748
2023-12-31    12242
2024-01-31     9458
2024-02-29     7528
2024-03-31     6309
2024-04-30     8530
2024-05-31     7379
2024-06-30     7864
2024-07-31     9115
2024-08-31     7010
2024-09-30     8821
2024-10-31    13907
2024-11-30    12989
2024-12-31    12346
Freq: ME, Name: quantidade, dtyp

In [ ]:
df.sort_values(['data', 'nota']).to_csv('/home/oscar/escola/dados/saida.csv', index=False)

# outros

In [228]:
d = {}
v = 4
q = 10000
for i in range(v):
  d[str(i)] = 0

for i in range(q):
  r = str(random.randrange(v))
  d[str(r)] += 1

print(d)

{'0': 2530, '1': 2494, '2': 2494, '3': 2482}


In [46]:
# df.query('horario < "2019-01-11"')
# df.loc[
#   df.horario < pd.to_datetime('2019-01-11', format='%Y-%m-%d') 
# ]

# h = df.horario
# h.mask(h < pd.to_datetime('2019-01-11', format='%Y-%m-%d') )
# h

# df.where(df.horario < pd.to_datetime('2019-01-11', format='%Y-%m-%d'))



id             horario   categoria                     produto  \
73538       NaN                 NaT         NaN                         NaN   
74345       NaN                 NaT         NaN                         NaN   
75692       NaN                 NaT         NaN                         NaN   
76343       NaN                 NaT         NaN                         NaN   
75699       NaN                 NaT         NaN                         NaN   
...         ...                 ...         ...                         ...   
39113  304165.0 2019-12-31 22:13:00   Batteries      AAA Batteries (4-pack)   
33863  299125.0 2019-12-31 22:21:00       Cable        USB-C Charging Cable   
40860  305840.0 2019-12-31 22:54:00  Headphones  Bose SoundSport Headphones   
35322  300519.0 2019-12-31 23:13:00  Headphones  Bose SoundSport Headphones   
35323  300519.0 2019-12-31 23:13:00       Cable    Lightning Charging Cable   

       quant  preco                                 endereco          cidade  
73538    NaN    NaN                                      NaN             NaN  
74345    NaN    NaN                                      NaN             NaN  
75692    NaN    NaN                                      NaN             NaN  
76343    NaN    NaN                                      NaN             NaN  
75699    NaN    NaN                                      NaN             NaN  
...      ...    ...                                      ...             ...  
39113    1.0   2.99         825 Adams St, Portland, OR 97035        Portland  
33863    1.0  11.95  754 Hickory St, New York City, NY 10001   New York City  
40860    1.0  99.99    784 River St, San Francisco, CA 94016   San Francisco  
35322    1.0  99.99   657 Spruce St, New York City, NY 10001   New York City  
35323    1.0  14.95   657 Spruce St, New York City, NY 10001   New York City  

[185950 rows x 8 columns]

In [81]:
# # Exemplo de DataFrame
# data = {'Frequencia': [10, 20, 30, 40, 50],
#         'Medicao': [15, 25, 35, 45, 55],
#         'Valor': [100, 200, 300, 400, 500]}

# df = pd.DataFrame(data)

# # Defina os níveis de frequência e medição
# nivel_frequencia = 30
# nivel_medicao = 40
# constante = 50

# # Use a função apply com uma função lambda para atualizar os valores
# df['Valor'] = df.apply(lambda row: row['Valor'] + constante if row['Frequencia'] > nivel_frequencia and row['Medicao'] > nivel_medicao else row['Valor'], axis=1)

# df



dfc = pd.DataFrame({
  'a': ['one', 'one', 'two', 'three', 'two', 'one', 'six'],
  'c': np.arange(7)})

mask = dfc['a'].str.startswith('o')

dfc.loc[mask, 'c'] = dfc['c'] + 42
dfc
# mask

a   c
0    one  42
1    one  43
2    two   2
3  three   3
4    two   4
5    one  47
6    six   6

In [203]:

population = [i + 1 for i in range(20)]
weights = [i * i + 1 for i in range(20, 0, -1)]
samples = random.choices(population, weights)[0]
samples

4

In [129]:
[i for i in range(20, 1, -1)]

[20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2]